fv3gfs/util/halo_updater.py

```diff
@@ -0,0 +1,263 @@
+                    packed_buffers, numpy_like_module, boundary
+                )
+                buffer.add_scalar_specification(
+                    specification,
+                    boundary.send_slice(specification, n_halo_points),
+                    boundary.n_clockwise_rotations,
+                    boundary.recv_slice(specification, n_halo_points),
+                )
+
+        for _to_rank, buffer in packed_buffers:
+            buffer.compile()
+
+        return cls(comm, tag, packed_buffers, timer)
+
+    @classmethod
+    def from_vector_specifications(
+        cls,
+        comm: "Communicator",
+        numpy_like_module: NumpyModule,
+        specifications_x: Iterable[HaloUpdateSpec],
+        specifications_y: Iterable[HaloUpdateSpec],
+        boundaries: Iterable[Boundary],
+        tag: int,
+        n_halo_points: int,
+        optional_timer: Optional[Timer] = None,
+    ) -> "HaloUpdater":
+        """Create/retrieve as many packed buffer as needed and queue the slices to exchange.
+
+        Args:
+            comm: communicator to post network messages
+            numpy_like_module: module implementing numpy API
+            quantities_x: quantities to exchange along the x axis.
+                          Length must match y quantities.
+            quantities_y: quantities to exchange along the y axis.
+                          Length must match x quantities.
+            boundaries: informations on the exchange boundaries.
+            tag: network tag (to differentiate messaging) for this node.
+            n_halo_points: size of the halo to exchange.
+            optional_timer: timing of operations.
+        
+        Returns:
+            HaloUpdater ready to exchange data.
+        """
+        timer = optional_timer if optional_timer is not None else NullTimer()
+
+        packed_buffers: List[Tuple[int, HaloDataTransformer]] = []
+        for boundary in boundaries:
+            for specification_x, specification_y in zip(
+                specifications_x, specifications_y
+            ):
+                buffer = cls._get_packed_buffer(
+                    packed_buffers, numpy_like_module, boundary
+                )
+                buffer.add_vector_specification(
+                    specification_x,
+                    specification_y,
+                    boundary.send_slice(specification_x, n_halo_points),
+                    boundary.send_slice(specification_y, n_halo_points),
+                    boundary.n_clockwise_rotations,
+                    boundary.recv_slice(specification_x, n_halo_points),
+                    boundary.recv_slice(specification_y, n_halo_points),
+                )
+
+        for _to_rank, buffer in packed_buffers:
+            buffer.compile()
+
+        return cls(comm, tag, packed_buffers, timer)
+
+    @classmethod
+    def _get_packed_buffer(
+        cls,
+        packed_buffer: List[Tuple[int, HaloDataTransformer]],
```

Comment:
`Dict[int, HaloDataTransformer]` keyed on `boundary.to_rank` would be more appropriate for this use case. If you use it you can reduce this function down to 3-5 lines because you no longer have to check for uniqueness or manually search a list for the key.

If you use a [DefaultDict](https://docs.python.org/3/library/collections.html#collections.defaultdict) that generates HaloDataTransformer, you could remove this class method altogether.

Reply:
Indeed this is a dict. I derived dict into it's own custom init dict to be able see `HaloTransformerDict`